embulk-input-s3/src/main/java/org/embulk/input/s3/S3FileInputUtils.java

```diff
@@ -0,0 +1,51 @@
+package org.embulk.input.s3;
+
+import org.embulk.spi.Exec;
+import org.embulk.spi.util.RetryExecutor;
+import org.slf4j.Logger;
+
+/**
+ * Utilit class for S3 File Input.
+ */
+public final class S3FileInputUtils
+{
+    private S3FileInputUtils()
+    {
+    }
+
+    public static final <T> T executeWithRetry(int maximumRetries, int initialRetryIntervalMillis, int maximumRetryIntervalMillis, AlwaysRetryRetryable<T> alwaysRetryRetryable)
+            throws RetryExecutor.RetryGiveupException, InterruptedException
+    {
+        return RetryExecutor.retryExecutor()
+                .withRetryLimit(maximumRetries)
+                .withInitialRetryWait(initialRetryIntervalMillis)
+                .withMaxRetryWait(maximumRetryIntervalMillis)
+                .runInterruptible(alwaysRetryRetryable);
+    }
+
+    public abstract static class AlwaysRetryRetryable<T> implements RetryExecutor.Retryable<T>
+    {
+        private static final Logger LOGGER = Exec.getLogger(AlwaysRetryRetryable.class);
+
+        @Override
+        public abstract T call() throws Exception;
+
+        @Override
+        public boolean isRetryableException(Exception exception)
+        {
+            return true;
+        }
+
+        @Override
+        public void onRetry(Exception exception, int retryCount, int retryLimit, int retryWait) throws RetryExecutor.RetryGiveupException
+        {
+            LOGGER.info("Retry [{}]/[{}] with retryWait [{}] on exception {}", retryCount, retryLimit, retryWait, exception.getMessage());
+        }
+
+        @Override
+        public void onGiveup(Exception firstException, Exception lastException) throws RetryExecutor.RetryGiveupException
+        {
+            LOGGER.info("Giving up execution on exception", lastException);
```

Comment:
Another nitpicking, Since first exception will automatically be re-thrown by RetryExecutor, I think this log should be more explicit it is logging out the "lastException", not the one that being rethrown (or it could log out both).